tests/agent_features/_test_async_generator_trace.py

```diff
@@ -0,0 +1,507 @@
+    ],
+)
+def test_async_generator_functions_outside_of_transaction(event_loop, trace):
+    @trace
+    async def agen():
+        for _ in range(2):
+            yield "foo"
+
+    async def _test():
+        assert [_ async for _ in agen()] == ["foo", "foo"]
+
+    event_loop.run_until_complete(_test())
+
+
+@validate_transaction_metrics(
+    "test_catching_generator_exit_causes_runtime_error",
+    background_task=True,
+    scoped_metrics=[("Function/agen", 1)],
+    rollup_metrics=[("Function/agen", 1)],
+)
+def test_catching_generator_exit_causes_runtime_error(event_loop):
+    @function_trace(name="agen")
+    async def agen():
+        try:
+            yield
+        except GeneratorExit:
+            yield
+
+    @background_task(name="test_catching_generator_exit_causes_runtime_error")
+    async def _test():
+        gen = agen()
+
+        # kickstart the coroutine (we're inside the try now)
+        await gen.asend(None)
+
+        # Generators cannot catch generator exit exceptions (which are injected by
+        # close). This will result in a runtime error.
+        with pytest.raises(RuntimeError):
+            await gen.aclose()
+
+        if is_pypy:
+            gen = None
+            gc.collect()
+
+    event_loop.run_until_complete(_test())
+
+
+@validate_transaction_metrics(
+    "test_async_generator_time_excludes_creation_time",
+    background_task=True,
+    scoped_metrics=[("Function/agen", 1)],
+    rollup_metrics=[("Function/agen", 1)],
+)
+def test_async_generator_time_excludes_creation_time(event_loop):
+    @function_trace(name="agen")
+    async def agen():
+        yield
+
+    metrics = []
+    full_metrics = {}
+
+    @capture_transaction_metrics(metrics, full_metrics)
+    @background_task(name="test_async_generator_time_excludes_creation_time")
+    def _test_async_generator_time_excludes_creation_time():
+        async def _test():
+            gen = agen()
+            time.sleep(0.1)
+            async for _ in gen:
+                pass
+
+        event_loop.run_until_complete(_test())
+    _test_async_generator_time_excludes_creation_time()
+
+    # check that the trace does not include the time between creation and
+    # consumption
+    assert full_metrics[("Function/agen", "")].total_call_time < 0.1
+
+
+@validate_transaction_metrics(
+    "test_complete_async_generator",
+    background_task=True,
+    scoped_metrics=[("Function/agen", 1)],
+    rollup_metrics=[("Function/agen", 1)],
+)
+@background_task(name="test_complete_async_generator")
+def test_complete_async_generator(event_loop):
+    @function_trace(name="agen")
+    async def agen():
+        for i in range(5):
+            yield i
+
+    async def _test():
+        gen = agen()
+        assert [x async for x in gen] == [x for x in range(5)]
+
+    event_loop.run_until_complete(_test())
+
+
+@pytest.mark.parametrize("nr_transaction", [True, False])
+def test_incomplete_async_generator(event_loop, nr_transaction):
```

Comment:
Couldn't this case create a runtime instrumentation error? If the generator doesn't complete, how do we ensure the call to `complete_root` would find an active trace?

Reply:
Surprisingly, when the transaction exits it closes any traces for async children automatically. I included a test for this case to demonstrate that it doesn't crash, and actually still reports a metric.